### DISCUSSION TITLE
结合项目经验存在一些问题辛苦解答

### DISCUSSION THREAD

**Discussion by BUPTlhuanyu** [galacean/effects-runtime#719] [Q&A] [answered]
1. 【销毁时机】看文档说是 loadScene 之后只能是合成对象有 end 事件，https://galacean.antgroup.com/effects/user/xqci1go27gptn33b，如果产物中有多个合成元素，如何知道整个动画结束了
2. 【如何销毁】直接调 dispose 报错，index.mjs:25092 Uncaught TypeError: Cannot read properties of undefined (reading 'meshes')
4. 【销毁的结果】页面中使用了 threejs 实现的游戏，使用的 webgl，并且页面里还有一些动效是使用 pag 实现的 https://pag.io/；动效播放完成之后，canvas 没有被删除，会有内存残留，同时 webgl context 也会到达上限
5. 【多个动效同时播放的稳定性问题】支持多个动效同时播放吗？
6. 【多个动效播放的内存问题】多个动效分时段播放可以复用一个 canvas 吗？
7. 【是否可以替代其他动效方案】pag 的优势是体积小，缺点是 web 的 lite 方案基于 video 的编解码，然后获取序列帧使用 canvas 绘制，处理流程过长，会导致一些莫名其妙的无法播放的问题，并且无法同时播放多个。目前这个方案是否可以替代 pag。和 lottie 相比这种动效方案有什么优势吗？
8. 【编辑器】编辑器开源吗？

**Comment by yiiqii** [comment, marked as answer]
1、一般在页面关闭时进行销毁（dispose）即可，不用在合成播放完成后手动 dispose。
1.2 如果产物中有多个合成元素，如何知道整个动画结束了？
  对所有合成的 end 做 Promise.all()
2、如果是在 composition end 回调直接 dispose 现在会这个错误（不过不会影响渲染），这个会在 2.1 版本优化掉
3、是使用的 galacean effects  的 threejs 插件吗？
4、支持的：https://galacean.antgroup.com/effects/playground/vanilla-%E5%A4%9A%E5%90%88%E6%88%90%E6%92%AD%E6%94%BE
5、可以，同上使用方式
6、是否可以替代其他方案不便直说，可以通过是否能直接满足你的业务诉求来判断即可。相比 lottie 优势有：1、基于 WebGL 渲染，较高的渲染性能；2、更丰富的效果，同时包含 2D、3D 表现形式
7、编辑器暂无开源计划

**Reply by BUPTlhuanyu** [reply]
还有个问题：动画的生命周期事件貌似文档里没有体现？还有就是播放进度的时间是否靠谱？底层是 requestAnimation 中基于绝对时间计算的播放进度吗？

**Reply by BUPTlhuanyu** [reply]
不是 galacean effects 的 threejs 插件，就是纯 threejs 实现的一个球类游戏

**Reply by BUPTlhuanyu** [reply]
第4个和第5个我表述有点问题，我想问的是，创建多个 player 同时播放，动效分散在页面的不同的地方，有一部分播放时间是重叠的这种

**Reply by yiiqii** [reply]
1、动画的生命周期事件 暂未暴露，2.1 会开放。播放进度是绝对的真实时间
2、那你指的 [canvas] 没有被删除是如何使用的？
3、一个页面不推荐多 player 同时播放，特殊场景（如：有弹窗）最多也 2 个 player，这种使用一般不会有性能问题。另外，一个 player 对应应该 canvas，不能共用

**Reply by BUPTlhuanyu** [reply]
![image](https://github.com/user-attachments/assets/6ac7c224-f2e8-4152-8b6c-c447bf5e55f7)

动画结束之后 webGL 还有残留，以及文字单独创建一个 canvas，元素多的话，canvas 数量也会比较多

**Reply by yiiqii** [reply]
player 执行 dispose 后，相关数据都会清理，
https://github.com/galacean/effects-runtime/blob/main/packages/effects/src/player.ts#L783
https://github.com/galacean/effects-runtime/blob/main/packages/effects-core/src/canvas-pool.ts#L10
如果浏览器没有及时清除，可能是浏览器自身的机制

**Reply by BUPTlhuanyu** [reply]
好的，多谢
